// select-kit/components/topic-notifications-button.js
import Component from "@glimmer/component";
import {
    tracked
} from "@glimmer/tracking";
import {
    hash
} from "@ember/helper";
import {
    action
} from "@ember/object";
import {
    service
} from "@ember/service";
import {
    htmlSafe
} from "@ember/template";
import {
    isEmpty
} from "@ember/utils";
import {
    NotificationLevels
} from "discourse/lib/notification-levels";
import i18n from "discourse-common/helpers/i18n";
import getURL from "discourse-common/lib/get-url";
import I18n from "discourse-i18n";
import TopicNotificationsOptions from "select-kit/components/topic-notifications-options";
import {
    precompileTemplate
} from "@ember/template-compilation";
import {
    setComponentTemplate
} from "@ember/component";
export default class TopicNotificationsButton extends Component {
    static {
        dt7948.g(this.prototype, "currentUser", [service]);
    }#
    currentUser = (dt7948.i(this, "currentUser"), void 0);
    static {
        dt7948.g(this.prototype, "isLoading", [tracked], function() {
            return false;
        });
    }#
    isLoading = (dt7948.i(this, "isLoading"), void 0);
    get notificationLevel() {
        return this.args.topic.get("details.notification_level");
    }
    get appendReason() {
        return this.args.appendReason ? ? true;
    }
    get showFullTitle() {
        return this.args.showFullTitle ? ? true;
    }
    get showCaret() {
        return this.args.showCaret ? ? true;
    }
    get reasonText() {
        const topic1 = this.args.topic;
        const level1 = topic1.get("details.notification_level") ? ? 1;
        const reason1 = topic1.get("details.notifications_reason_id");
        let localeString1 = `topic.notifications.reasons.${level1}`;
        if (typeof reason1 === "number") {
            let localeStringWithReason1 = `${localeString1}_${reason1}`;
            if (this._reasonStale(level1, reason1)) {
                localeStringWithReason1 += "_stale";
            }
            // some sane protection for missing translations of edge cases
            if (I18n.lookup(localeStringWithReason1, {
                    locale: "en"
                })) {
                localeString1 = localeStringWithReason1;
            }
        }
        if (this.currentUser ? .user_option.mailing_list_mode && level1 > NotificationLevels.MUTED) {
            return I18n.t("topic.notifications.reasons.mailing_list_mode");
        } else {
            return I18n.t(localeString1, {
                username: this.currentUser ? .username_lower,
                basePath: getURL("")
            });
        }
    }
    // The user may have changed their category or tag tracking settings
    // since this topic was tracked/watched based on those settings in the
    // past. In that case we need to alter the reason message we show them
    // otherwise it is very confusing for the end user to be told they are
    // tracking a topic because of a category, when they are no longer tracking
    // that category.
    _reasonStale(level1, reason1) {
        if (!this.currentUser) {
            return;
        }
        const watchedCategoryIds1 = this.currentUser.watched_category_ids || [];
        const trackedCategoryIds1 = this.currentUser.tracked_category_ids || [];
        const watchedTags1 = this.currentUser.watched_tags || [];
        if (this.args.topic.category_id) {
            if (level1 === 2 && reason1 === 8) {
                // 2_8 tracking category
                return !trackedCategoryIds1.includes(this.args.topic.category_id);
            } else if (level1 === 3 && reason1 === 6) {
                // 3_6 watching category
                return !watchedCategoryIds1.includes(this.args.topic.category_id);
            }
        } else if (!isEmpty(this.args.topic.tags)) {
            if (level1 === 3 && reason1 === 10) {
                // 3_10 watching tag
                return !this.args.topic.tags.some(tag1 => watchedTags1.includes(tag1));
            }
        }
        return false;
    }
    async changeTopicNotificationLevel(levelId1) {
        if (levelId1 === this.notificationLevel) {
            return;
        }
        this.isLoading = true;
        try {
            await this.args.topic.details.updateNotifications(levelId1);
        } finally {
            this.isLoading = false;
        }
    }
    static {
        dt7948.n(this.prototype, "changeTopicNotificationLevel", [action]);
    }
    static {
        setComponentTemplate(precompileTemplate("\n    <div class=\"topic-notifications-button\">\n      {{#if this.appendReason}}\n        <p class=\"reason\">\n          <TopicNotificationsOptions @value={{this.notificationLevel}} @topic={{@topic}} @onChange={{this.changeTopicNotificationLevel}} @options={{hash icon=(if this.isLoading \"spinner\") showFullTitle=this.showFullTitle showCaret=this.showCaret headerAriaLabel=(i18n \"topic.notifications.title\")}} />\n          <span class=\"text\">{{htmlSafe this.reasonText}}</span>\n        </p>\n      {{else}}\n        <TopicNotificationsOptions @value={{this.notificationLevel}} @topic={{@topic}} @onChange={{this.changeTopicNotificationLevel}} @options={{hash icon=(if this.isLoading \"spinner\") showFullTitle=this.showFullTitle showCaret=this.showCaret headerAriaLabel=(i18n \"topic.notifications.title\")}} />\n      {{/if}}\n    </div>\n  ", {
            strictMode: true,
            scope: () => ({
                TopicNotificationsOptions,
                hash,
                i18n,
                htmlSafe
            })
        }), this);
    }
}
