// float-kit/modifiers/close-on-escape.js
import {
    registerDestructor
} from "@ember/destroyable";
import {
    service
} from "@ember/service";
import Modifier from "ember-modifier";
import {
    bind
} from "discourse-common/utils/decorators";
export default class FloatKitCloseOnEscape extends Modifier {
    static {
        dt7948.g(this.prototype, "menu", [service]);
    }#
    menu = (dt7948.i(this, "menu"), void 0);
    constructor(owner, args) {
        super(owner, args);
        registerDestructor(this, instance => instance.cleanup());
    }
    modify(element, [closeFn]) {
        this.closeFn = closeFn;
        this.element = element;
        document.addEventListener("keydown", this.check);
    }
    check(event) {
        if (event.key === "Escape") {
            event.stopPropagation();
            event.preventDefault();
            this.closeFn();
        }
    }
    static {
        dt7948.n(this.prototype, "check", [bind]);
    }
    cleanup() {
        document.removeEventListener("keydown", this.check);
    }
}
